Add user id, license, and bbox arguments to photo search function

elipousson opened the pull request:
I noticed that @fozy81 recently added a photo search function. I expanded on the function to allow user id, license, or geographical bounding box as search arguments. I also added the option to return extra fields. The extra fields optionally include direct link to the photo files so it could also help to implement #6.

Hope this is a welcome addition to the package!

elipousson committed:
Allow additional optional arguments for getPhotoSearch function

Add flickr.photos.search arguments for user_id, licence_id, and bbox.
Optionally returns extra fields.
Change per page default to 100 (max for bbox search is 250).

elipousson committed:
Update list of getPhotoSearch parameters in Package Contents description

elipousson committed:
Replace print with stop for extra fields error message

fozy81 commented:
Wow, lots of great improvements @elipousson :+1: 
I tested out the code and works fine for me :fireworks:  I would maybe suggest we could reduce the number of  `if` statements by setting default `NULL` parameters to the arguments. For example, if the url includes `user_id=`  and the user_id is blank/null it still works.  e.g. ...`user_id=&tags=cats&bbox=1.73,-74.1,40.01,-7.12&per_page=100&page=1&format=json&nojsoncallback=1`
So I think we can:
```
getPhotoSearch <- function(api_key,
                           user_id = NULL,
                           tags = NULL,
                           licence_id,
                           bbox = NULL,
                           extras,
                           per_page = 100,
                           page = 1)
```
And then remove the if statements. leaving just `url = paste0(url,  "&user_id=", stringr::str_replace(user_id, "@", "%40")` etc - same for `bbox`. But really, doesn't make much difference :shrug: . The `extra_fields` + licence name checks is really useful feature so no change there.  
My other thoughts are maybe add to vignette and/or expand the examples to include details on how to use these extra arguments. But just my comments, over to @koki25ando for review.

elipousson committed:
Remove checks by setting null defaults + add sort parameter

Add examples to illustrate user_id, sort, extras, bbox parameters

elipousson committed:
Add null default value for license ID parameter

elipousson commented:
Thanks for the suggestions, @fozy81. I added the null defaults as suggested and simplified the license name check into a single if statement. I added in a few more examples to show how the new features work. I also implemented the sort parameter feature from the Flickr photo search API which works with a similar check to the license check.

One small issue I noticed: including the description in the extra fields can result in this error—"Error: lexical error: invalid character inside string." I don't know if it is possible to sanitize the description text or make it fail more gracefully and drop the description if it won't work?

fozy81 commented:
Issue sounds familiar... :thinking: ...You could try  jsonlite::validate(description) to check if description is valid if not then exclude. 

But there must be a way to decode/encode the description somehow? Not sure of the best way to do this. Do you have a example of the error you could share (excluding your APi key!)?

elipousson commented:
This search returns the error: 

`getPhotoSearch(api_key = "XXXX", user_id = "grand_canyon_nps", extras = c("description"))`

It definitely depends on the content of the description because the same search works for other user accounts without encountering the error.

koki25ando commented on R/getPhotoSearch.R:
```diff
@@ -1,28 +1,103 @@
+#' R access to flickr search api
 #'
-#' @param api_key Your API application key
+#' @param api_key Required. Your API application key
+#' @param user_id The NSID of the user who's photo to search. If this parameter isn't passed then everybody's public photos will be searched.
 #' @param tags List of tags you wish to search for e.g. c("cats, "dogs)
+#' @param license_id The license id for photos (for possible values see the Flickr API method flickr.photos.licenses.getInfo).
+#' @param sort Order to sort returned photos. The possible values are: "date-posted-asc", "date-posted-desc", "date-taken-asc", "date-taken-desc", "interestingness-desc", "interestingness-asc", and "relevance"
+#' @param bbox A object of class bbox or a comma-delimited string of 4 values defining the Bounding Box of the area that will be searched. The 4 values represent the bottom-left corner of the box and the top-right corner, minimum_longitude, minimum_latitude, maximum_longitude, maximum_latitude.
+#' @param extras A vector of extra information to fetch for each returned record. Currently supported fields are: c("description", "license", "date_upload", "date_taken", "owner_name", "icon_server", "original_format", "last_update", "geo", "tags", "machine_tags", "o_dims", "views", "media", "path_alias", "url_sq", "url_t", "url_s", "url_q", "url_m", "url_n", "url_z", "url_c", "url_l", "url_o")
+#' @param per_page Number specifying how many results per page to return. Default 100 results per page. Maximum of 250 if bbox provided or 500 otherwise.
```
really cool features!!!! thanks for adding

koki25ando commented on R/getPhotoSearch.R:
```diff
@@ -1,28 +1,103 @@
+
+  url = paste0(url, "&sort=", sort)
+
+  if (class(bbox) == "bbox") {
+    bbox <- paste0(bbox[1], ",", bbox[2], ",", bbox[3], ",", bbox[4])
+  }
+
+  url = paste0(url, "&bbox=", bbox)
```
selecting areas is really nice feature!!!!! i love it👍

but i actually encountered some errors when i set parameteres with spaces. here is the code which returned an error
```
getPhotoSearch(api_key = API_KEY,
               tags = "panda",
               bbox = "139.7682226529, 35.712627977, 139.7724605432, 35.7181464141",
               extras = c("geo", "owner_name", "tags"))
```
> error: lexical error: invalid char in json text.
                                       <!DOCTYPE HTML PUBLIC "-//IETF/
                     (right here) ------^ `

so here is my suggestion to remove all the scpaces from bbox input string

```url = paste0(url, "&bbox=", stringr::str_remove_all(bbox, "\\s"))```

elipousson replied:
On reflection, I wonder if it makes sense to require a vector of coordinates instead of allowing strings at all. This examples would be changed to look like:

`getPhotoSearch(api_key = API_KEY,
               tags = "panda",
               bbox = c(139.7682226529, 35.712627977, 139.7724605432, 35.7181464141),
               extras = c("geo", "owner_name", "tags"))`

The same code used to turn a bbox class object into the appropriate string for the Flickr API call would work for this vector xmin, ymin, xmax, ymax format just as well.

koki25ando reviewed (commented):
guys, thank you for sharing!!! lots of nice improvements!!
Your contributions are always appreciated!!!!!
i will check the scripts deeper but here are some things i  noticed.

fozy81 commented:
Hi @elipousson & @koki25ando 

I  tried adding the `.mapUnicode = FALSE` argument to `getURL` and this worked! Like so: `raw_data <- RCurl::getURL(url, ssl.verifypeer = FALSE, .mapUnicode = FALSE)`

I think by default `getURL` is trying to map codes into representative unicode icons. For instance `\u263A` into the :smile:  icon. But because there is html in this particular description field it is getting confused(?).  So using `.mapUnicode = FALSE` means the raw description is returned - I think this is okay? My knowledge of unicodes and text parsing is limited! 

:+1:  for vector for coordinates.

elipousson committed:
Modify requirements for bbox parameter + change getURL function to mapUnicode = FALSE

Fix example of bbox search to show new requirement for numeric vector

elipousson commented:
Great! I think that last commit should make it easier for someone to set the bbox correctly than trying to pass it as a character string, @koki25ando. I also added the `.mapUnicode = FALSE` parameter – thanks for figuring that our @fozy81. There are a few more features from the full search API that could be fun to add in the future but I think this is good to go.

koki25ando commented:
@elipousson looks good👍 thanks for your commit!
